nucliadb_utils/nucliadb_utils/nats/demux.py

```diff
@@ -0,0 +1,310 @@
+        logger.debug(
+            "Message will be soon demultiplexed and processed",
+            extra={"subject": subject, "reply": reply, "seqid": seqid},
+        )
+        asyncio.create_task(self.safe_demux(msg))
+
+    async def safe_demux(self, msg: Msg):
+        try:
+            await self.demux(msg)
+        except Exception as exc:
+            subject = msg.subject
+            reply = msg.reply
+            seqid = int(msg.reply.split(".")[5])
+
+            event_id = capture_exception(exc)
+            logger.error(
+                "Error while concurrently handling a NATS message. "
+                f"Check sentry for more details. Event id: {event_id}",
+                extra={
+                    "subject": subject,
+                    "reply": reply,
+                    "seqid": seqid,
+                },
+                exc_info=exc,
+            )
+
+            await msg.nak()
+            raise exc
+
+    async def demux(self, msg: Msg):
+        async with MessageProgressUpdater(
+            msg, nats_consumer_settings.nats_ack_wait * 0.66
+        ):
+            subject = msg.subject
+            reply = msg.reply
+            seqid = int(msg.reply.split(".")[5])
+
+            split, work = self.processor.splitter(msg)
+            logger.info(
+                f"Message demultiplexed to split '{split}'",
+                extra={"subject": subject, "reply": reply, "seqid": seqid},
+            )
+
+            worker, created = self.get_or_create_worker(split)
+
+            seqid = int(msg.reply.split(".")[5])
+            work_id = f"{split}:{seqid}"
+            work_unit = WorkUnit(work_id, work)
+
+            if work_id in self.working_on:
+                logger.warning(
+                    f"Trying to process again an already received NATS message. Skipping (workid={work_id})"
+                )
+                return
+
+            await worker.add_work(work_unit)
+            self.working_on.add(work_id)
+
+            if created or (not created and not worker.working):
+                # The second condition should never happen, the worker task finished
+                # between us getting the worker and adding work. We'll create a new
+                # task for it
+                self.start_worker(split)
+
+            loop = asyncio.get_event_loop()
+            wait_condition = loop.create_future()
+
+            work_unit_done.add_listener(
+                work_id,
+                partial(
+                    NatsDemultiplexer.on_work_unit_done,
+                    work_id=work_id,
+                    notifier=wait_condition,
+                ),
+            )
+
+            work_outcome = await wait_condition
+            self.working_on.discard(work_id)
+            if work_outcome.error:
+                await msg.nak()
+            else:
+                if work_outcome.success:
+                    await msg.ack()
+                else:
+                    await msg.nak()
```

Comment:
though: do we need more logging in this method overall? To have evidence that things have been acked or nacked and why.